### PR TITLE
Initial LoongArch support

### PR DESCRIPTION
**Pull request by FlyGoat**
Hi All,

This PR implemented initial support for LoongArch.
Currently, only LoongArch 64 build is supported as upstream LoongArch32 ABI is not ratified yet, but all implementations are 32bit aware and build tested against downstream toolchain.

It requires a custom `loongarch64-unknown-elf` toolchain to build, my `ct-ng` config built for current crosstool-ng master:
```
CT_CONFIG_VERSION="4"
CT_EXPERIMENTAL=y
CT_ARCH_LOONGARCH=y
# CT_DEMULTILIB is not set
CT_ARCH_USE_MMU=y
CT_ARCH_ARCH="loongarch64"
CT_LIBC_NONE=y
CT_CC_GCC_LIBSTDCXX_VERBOSE=m
CT_CC_LANG_CXX=y
CT_DEBUG_GDB=y
# CT_GDB_CROSS_PYTHON is not set
CT_GETTEXT_NEEDED=y
```

For QEMU testing, a patch to enable semihosting is also required:
https://lore.kernel.org/qemu-devel/20241222-semihosting-v1-1-8a770df60e9c@flygoat.com/

All tests are passing on my side:
```
ninja test
[0/1] Running all tests.
  1/206 picolibc:headers / check-cdef-semihost.h                    OK               0.05s
  2/206 picolibc:headers / check-cdef-stdio.h                       OK               0.05s
  3/206 picolibc:headers / check-cdef-stdio-bufio.h                 OK               0.05s
  4/206 picolibc:headers / check-cdef-sys-auxv.h                    OK               0.05s
  5/206 picolibc:headers / check-cdef-sys-cdefs.h                   OK               0.05s
  6/206 picolibc:headers / check-cdef-sys-errno.h                   OK               0.04s
  7/206 picolibc:headers / check-cdef-sys-fcntl.h                   OK               0.04s
  8/206 picolibc:headers / check-cdef-sys-file.h                    OK               0.04s
  9/206 picolibc:headers / check-cdef-sys-iconvnls.h                OK               0.03s
 10/206 picolibc:headers / check-cdef-sys-lock.h                    OK               0.03s
 11/206 picolibc:headers / check-cdef-sys-param.h                   OK               0.03s
 12/206 picolibc:headers / check-cdef-sys-queue.h                   OK               0.03s
 13/206 picolibc:headers / check-cdef-sys-dir.h                     OK               0.09s
 14/206 picolibc:headers / check-cdef-sys-resource.h                OK               0.07s
 15/206 picolibc:headers / check-cdef-sys-sched.h                   OK               0.07s
 16/206 picolibc:headers / check-cdef-sys-select.h                  OK               0.08s
 17/206 picolibc:headers / check-cdef-sys-stat.h                    OK               0.07s
 18/206 picolibc:headers / check-cdef-sys-syslimits.h               OK               0.07s
 19/206 picolibc:headers / check-cdef-sys-timeb.h                   OK               0.06s
 20/206 picolibc:headers / check-cdef-sys-time.h                    OK               0.06s
 21/206 picolibc:headers / check-cdef-sys-times.h                   OK               0.06s
 22/206 picolibc:headers / check-cdef-sys-timespec.h                OK               0.06s
 23/206 picolibc:headers / check-cdef-sys-types.h                   OK               0.05s
 24/206 picolibc:headers / check-cdef-sys-unistd.h                  OK               0.05s
 25/206 picolibc:headers / check-cdef-sys-utime.h                   OK               0.05s
 26/206 picolibc:headers / check-cdef-sys-wait.h                    OK               0.04s
 27/206 picolibc:headers / check-cdef-rpc-types.h                   OK               0.04s
 28/206 picolibc:headers / check-cdef-sys-tree.h                    OK               0.09s
 29/206 picolibc:headers / check-cdef-rpc-xdr.h                     OK               0.07s
 30/206 picolibc:headers / check-cdef-arpa-inet.h                   OK               0.08s
 31/206 picolibc:headers / check-cdef-alloca.h                      OK               0.07s
 32/206 picolibc:headers / check-cdef-argz.h                        OK               0.07s
 33/206 picolibc:headers / check-cdef-ar.h                          OK               0.07s
 34/206 picolibc:headers / check-cdef-assert.h                      OK               0.06s
 35/206 picolibc:headers / check-cdef-byteswap.h                    OK               0.06s
 36/206 picolibc:headers / check-cdef-ctype.h                       OK               0.06s
 37/206 picolibc:headers / check-cdef-devctl.h                      OK               0.05s
 38/206 picolibc:headers / check-cdef-elf.h                         OK               0.05s
 39/206 picolibc:headers / check-cdef-endian.h                      OK               0.05s
 40/206 picolibc:headers / check-cdef-envlock.h                     OK               0.04s
 41/206 picolibc:headers / check-cdef-envz.h                        OK               0.04s
 42/206 picolibc:headers / check-cdef-errno.h                       OK               0.04s
 43/206 picolibc:headers / check-cdef-dirent.h                      OK               0.09s
 44/206 picolibc:headers / check-cdef-fastmath.h                    OK               0.07s
 45/206 picolibc:headers / check-cdef-fcntl.h                       OK               0.06s
 46/206 picolibc:headers / check-cdef-fenv.h                        OK               0.07s
 47/206 picolibc:headers / check-cdef-fnmatch.h                     OK               0.07s
 48/206 picolibc:headers / check-cdef-getopt.h                      OK               0.07s
 49/206 picolibc:headers / check-cdef-glob.h                        OK               0.07s
 50/206 picolibc:headers / check-cdef-grp.h                         OK               0.07s
 51/206 picolibc:headers / check-cdef-iconv.h                       OK               0.06s
 52/206 picolibc:headers / check-cdef-ieeefp.h                      OK               0.06s
 53/206 picolibc:headers / check-cdef-inttypes.h                    OK               0.06s
 54/206 picolibc:headers / check-cdef-langinfo.h                    OK               0.06s
 55/206 picolibc:headers / check-cdef-libgen.h                      OK               0.05s
 56/206 picolibc:headers / check-cdef-limits.h                      OK               0.05s
 57/206 picolibc:headers / check-cdef-locale.h                      OK               0.05s
 58/206 picolibc:headers / check-cdef-malloc.h                      OK               0.05s
 59/206 picolibc:headers / check-cdef-math.h                        OK               0.07s
 60/206 picolibc:headers / check-cdef-memory.h                      OK               0.07s
 61/206 picolibc:headers / check-cdef-ndbm.h                        OK               0.08s
 62/206 picolibc:headers / check-cdef-pwd.h                         OK               0.08s
 63/206 picolibc:headers / check-cdef-regex.h                       OK               0.08s
 64/206 picolibc:headers / check-cdef-sched.h                       OK               0.08s
 65/206 picolibc:headers / check-cdef-search.h                      OK               0.08s
 66/206 picolibc:headers / check-cdef-setjmp.h                      OK               0.07s
 67/206 picolibc:headers / check-cdef-signal.h                      OK               0.07s
 68/206 picolibc:headers / check-cdef-spawn.h                       OK               0.07s
 69/206 picolibc:headers / check-cdef-stdint.h                      OK               0.07s
 70/206 picolibc:headers / check-cdef-stdnoreturn.h                 OK               0.06s
 71/206 picolibc:headers / check-cdef-stdlib.h                      OK               0.06s
 72/206 picolibc:headers / check-cdef-string.h                      OK               0.06s
 73/206 picolibc:headers / check-cdef-strings.h                     OK               0.05s
 74/206 picolibc:headers / check-cdef-threads.h                     OK               0.09s
 75/206 picolibc:headers / check-cdef-time.h                        OK               0.08s
 76/206 picolibc:headers / check-cdef-uchar.h                       OK               0.08s
 77/206 picolibc:headers / check-cdef-unctrl.h                      OK               0.09s
 78/206 picolibc:headers / check-cdef-unistd.h                      OK               0.09s
 79/206 picolibc:headers / check-cdef-utime.h                       OK               0.09s
 80/206 picolibc:headers / check-cdef-wchar.h                       OK               0.09s
 81/206 picolibc:headers / check-cdef-wctype.h                      OK               0.09s
 82/206 picolibc:headers / check-cdef-wordexp.h                     OK               0.09s
 83/206 picolibc:headers / check-cdef-complex.h                     OK               0.09s
 84/206 picolibc:headers / check-cdef-picotls.h                     OK               0.08s
 85/206 picolibc / iconvjp                                          OK               0.14s
 86/206 picolibc / iconvru                                          OK               0.14s
 87/206 picolibc / hsearchtest                                      OK               0.16s
 88/206 picolibc / atexit                                           OK               0.18s
 89/206 picolibc / nulprintf                                        OK               0.19s
 90/206 picolibc / iconvnm                                          OK               0.21s
 91/206 picolibc / size_max                                         OK               0.21s
 92/206 picolibc / memcpy-1                                         OK               0.20s
 93/206 picolibc / memmove1                                         OK               0.23s
 94/206 picolibc / tzset                                            OK               0.21s
 95/206 picolibc / twctrans                                         OK               0.18s
 96/206 picolibc / tstring                                          OK               0.23s
 97/206 picolibc / asctime                                          OK               0.20s
 98/206 picolibc / tiswctype                                        OK               0.23s
 99/206 picolibc / printf_scanf                                     OK               0.19s
100/206 picolibc / printff_scanff                                   OK               0.18s
101/206 picolibc / printfi_scanfi                                   OK               0.16s
102/206 picolibc / printfm_scanfm                                   OK               0.15s
103/206 picolibc / printf-tests                                     OK               0.18s
104/206 picolibc / printfl_scanfl                                   OK               0.25s
105/206 picolibc / printff-tests                                    OK               0.18s
106/206 picolibc / printfi-tests                                    OK               0.17s
107/206 picolibc / printfm-tests                                    OK               0.15s
108/206 picolibc / time-sprintf                                     OK               0.17s
109/206 picolibc / hosted-exit                                      OK               0.15s
110/206 picolibc / hosted-exit-fail                                 EXPECTEDFAIL     0.15s   exit status 1
111/206 picolibc / try-ilp32                                        OK               0.20s
112/206 picolibc / math_test                                        OK               0.37s
113/206 picolibc / abort                                            EXPECTEDFAIL     0.16s   (exit status 134 or signal 6 SIGABRT)
114/206 picolibc / constructor-skip                                 OK               0.15s
115/206 picolibc / rounding-mode                                    OK               0.16s
116/206 picolibc / math_errhandling                                 OK               0.19s
117/206 picolibc / test-except                                      SKIP             0.16s   exit status 77
118/206 picolibc / ungetc                                           OK               0.14s
119/206 picolibc / regex                                            OK               0.15s
120/206 picolibc / malloc                                           OK               0.14s
121/206 picolibc / test-ubsan                                       OK               0.18s
122/206 picolibc / atexit                                           OK               0.15s
123/206 picolibc / setjmp                                           OK               0.18s
124/206 picolibc / on_exit                                          OK               0.17s
125/206 picolibc / long_double                                      OK               0.26s
126/206 picolibc / ffs                                              OK               0.22s
127/206 picolibc / timegm                                           OK               0.19s
128/206 picolibc / test-strchr                                      OK               0.20s
129/206 picolibc / test-raise                                       OK               0.18s
130/206 picolibc / test-sprintf-percent-n                           SKIP             0.15s   exit status 77
131/206 picolibc / test-put                                         OK               0.21s
132/206 picolibc / test-ctype                                       OK               0.18s
133/206 picolibc / test-uchar                                       OK               0.18s
134/206 picolibc / constructor                                      OK               0.18s
135/206 picolibc / test-vfscanf-percent-a                           OK               0.16s
136/206 picolibc / posix-io                                         OK               0.15s
137/206 picolibc / test-atomic                                      OK               0.17s
138/206 picolibc / test-fopen                                       OK               0.16s
139/206 picolibc / test-mktemp                                      OK               0.16s
140/206 picolibc / test-tmpnam                                      OK               0.18s
141/206 picolibc / t_fmemopen                                       OK               0.29s
142/206 picolibc / test-ungetc-ftell                                OK               0.19s
143/206 picolibc / test-wchar                                       OK               0.16s
144/206 picolibc / test-fgetc                                       OK               0.19s
145/206 picolibc / test-fgets-eof                                   OK               0.18s
146/206 picolibc / strcmp-1                                         OK               0.87s
147/206 picolibc / fenv                                             OK               0.19s
148/206 picolibc / rand                                             OK               0.18s
149/206 picolibc / test-efcvt                                       OK               0.17s
150/206 picolibc / test-strtod                                      OK               0.15s
151/206 picolibc / complex-funcs                                    OK               0.18s
152/206 picolibc / test-fma                                         OK               0.21s
153/206 picolibc / test-funopen                                     OK               0.13s
154/206 picolibc / time-tests                                       OK               0.13s
155/206 picolibc / tls                                              OK               0.13s
156/206 picolibc / math-funcs                                       OK               0.18s
157/206 picolibc / test-memset                                      OK               0.56s
158/206 picolibc / test-cplusplus                                   OK               0.03s
159/206 picolibc / test-memset_s                                    OK               0.15s
160/206 picolibc / test-memcpy_s                                    OK               0.16s
161/206 picolibc / test-memmove_s                                   OK               0.17s
162/206 picolibc / test-strcpy_s                                    OK               0.15s
163/206 picolibc / test-strerror_s                                  OK               0.16s
164/206 picolibc / test-strcat_s                                    OK               0.19s
165/206 picolibc / test-strerrorlen_s                               OK               0.17s
166/206 picolibc / test-strncat_s                                   OK               0.18s
167/206 picolibc / test-strnlen_s                                   OK               0.16s
168/206 picolibc / test-strncpy_s                                   OK               0.20s
169/206 picolibc / stack-smash                                      OK               0.17s
170/206 picolibc / test-sprintf_s                                   OK               0.18s
171/206 picolibc / dirname                                          OK               0.16s
172/206 picolibc / basename                                         OK               0.17s
173/206 picolibc / fnmatch                                          OK               0.19s
174/206 picolibc / qsort                                            OK               0.19s
175/206 picolibc / snprintf                                         OK               0.20s
176/206 picolibc / strtol                                           OK               0.16s
177/206 picolibc / sscanf                                           OK               0.21s
178/206 picolibc / string                                           OK               0.21s
179/206 picolibc:semihost / semihost-argv                           OK               0.17s
180/206 picolibc / strtod                                           OK               0.21s
181/206 picolibc:semihost / semihost-exit                           OK               0.19s
182/206 picolibc:semihost / semihost-errno                          OK               0.17s
183/206 picolibc:semihost / semihost-clock                          OK               0.18s
184/206 picolibc:semihost / semihost-exit-extended                  OK               0.19s
185/206 picolibc:semihost / semihost-flen                           OK               0.20s
186/206 picolibc:semihost / semihost-gettimeofday                   OK               0.17s
187/206 picolibc:semihost / semihost-get-cmdline                    OK               0.19s
188/206 picolibc:semihost / semihost-istty                          OK               0.16s
189/206 picolibc:semihost / semihost-open                           OK               0.18s
190/206 picolibc:semihost / semihost-read                           OK               0.18s
191/206 picolibc:semihost / semihost-remove                         OK               0.17s
192/206 picolibc:semihost / semihost-rename                         OK               0.16s
193/206 picolibc:semihost / semihost-seek                           OK               0.17s
194/206 picolibc:semihost / semihost-time                           OK               0.15s
195/206 picolibc:semihost / semihost-system                         OK               0.18s
196/206 picolibc:semihost / semihost-times                          OK               0.16s
197/206 picolibc:semihost / semihost-write0                         OK               0.13s
198/206 picolibc:semihost / semihost-writec                         OK               0.16s
199/206 picolibc:semihost / semihost-exit-failure                   EXPECTEDFAIL     0.14s   exit status 1
200/206 picolibc:semihost / semihost-exit-extended-failure          EXPECTEDFAIL     0.13s   exit status 1
201/206 picolibc:semihost / semihost-system-failure                 EXPECTEDFAIL     0.13s   exit status 1
202/206 picolibc:semihost / semihost-no-argv                        OK               0.13s
203/206 picolibc / malloc_stress                                    OK               0.81s
204/206 picolibc / test-fread-fwrite                                OK               1.09s
205/206 picolibc / test-long-longl                                  OK               1.42s
206/206 picolibc / test-long-long                                   OK               1.42s

Ok:                 199 
Expected Fail:      5   
Fail:               0   
Unexpected Pass:    0   
Skipped:            2   
Timeout:            0   
```

### CODE REVIEW

**Comment by keith-packard** [comment]
This all looks awesome! We'll need a compiler and qemu instance that we can incorporate into CI for testing before getting this merged though. That means making tarballs that run on Debian unstable (which will run essentially any Linux binaries) that CI can occasionally download. CI caches the container images, so new binaries are only downloaded when the CI configuration changes. Let me know if you can't find a place to host them and we can dump them on a server I've got at OSU OSL.

**Comment by FlyGoat** [comment]
@keith-packard Thanks for the feedback! I attempted to get CI jobs working on my side but found it frequently have problems on pulling artefacts from maps.altusmetrum.org, sometimes it's certification problem, sometimes it's timeout in middle...

```
2024-12-27T22:09:55.3138123Z #7 111.1 --2024-12-27 22:09:55--  https://maps.altusmetrum.org/m68k-unknown-elf.tar.xz
2024-12-27T22:09:55.6327301Z #7 111.1 Resolving maps.altusmetrum.org (maps.altusmetrum.org)... 140.211.9.9
2024-12-27T22:09:55.7342739Z #7 111.4 Connecting to maps.altusmetrum.org (maps.altusmetrum.org)|140.211.9.9|:443... connected.
2024-12-27T22:09:55.7343980Z #7 111.5 ERROR: The certificate of 'maps.altusmetrum.org' is not trusted.
2024-12-27T22:09:55.7344566Z #7 111.5 ERROR: The certificate of 'maps.altusmetrum.org' has expired.
2024-12-27T22:09:55.7345840Z #7 111.5 The certificate has expired
```

So I attempted to reproduce tools used here with github action and delivering artefacts with github release, this resulted side project: https://github.com/FlyGoat/picolibc-ci-tools, which build toolchains and qemu automatically and upload to github release on any github tag.

I wonder if it's possible to pull this project to picolibc organisation on github, I think it can ease pressure on your server, as well as improve reproducibility of CI jobs.

**Comment by FlyGoat** [comment]
I'm not really sure if there are any modification to QEMU source in `qemu-arm` and `qemu-m68k` so I didn't build them for qemu-stable, but it should be straightforward to add extra targets or even QEMU variants.

Also, I'm planning to get lm32 and or1k build tested (for litex project), so I built toolchains for them as well.

**Comment by keith-packard** [comment]
> So I attempted to reproduce tools used here with github action and delivering artefacts with github release, this resulted side project: https://github.com/FlyGoat/picolibc-ci-tools, which build toolchains and qemu automatically and upload to github release on any github tag.

What a good plan. I've done something temporary to work around this issue -- I've uploaded all of the existing tarballs to https://github.com/picolibc/toolchains so that we can at least get things working. Your plan of actually building them using github CI is better, but that will take some time to make sure we get the right bits built with the right options. Most of those bits have local bug fixes, some of which haven't been released upstream yet. All of this is in PR #899 

> I wonder if it's possible to pull this project to picolibc organisation on github, I think it can ease pressure on your server, as well as improve reproducibility of CI jobs.

Yes, absolutely. We can transition away from the above hack to auto-built toolchains as we get them working. For now,
let's just get your loongarch qemu bits split into a separate file so we can add that to the CI without affecting other qemu bits there.

**Comment by FlyGoat** [comment]
Currently only qemu-system-loongarch64 is built into qemu-stable tarball :-)

I’m waiting CI to pass in my tree: https://github.com/FlyGoat/picolibc/actions/runs/12521217442

**Comment by keith-packard** [comment]
ok, I've fixed CI so it pulls toolchain bits from github now. If you'll rebase, I can let CI run on the result and you should be able to get things tested and ready for integration.

**Comment by FlyGoat** [comment]
@keith-packard Thanks! done.

**Review by keith-packard — CHANGES_REQUESTED** [review]
couple of fixes needed, but looks really nice

File: newlib/libc/machine/loongarch/machine/_fpmath.h [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+/*
+ * Change unsigned int/long used by FreeBSD to fixed width types because
+ * ilp32 has a different size for unsigned long. --joel (20 Aug 2022)
+ */
+#include <stdint.h>
+
+union IEEEl2bits {
+	long double	e;
+	struct {
+		uint64_t	manl	:64;
```
[the remark this answers is not in the document]
Comment (FlyGoat) [reply newlib/libc/machine/loongarch/machine/_fpmath.h:40]: I just blindly copied that from RISC-V, AFAIK there is no MSB mode, should I remove it?

File: newlib/libc/machine/loongarch/machine/fenv-fp.h [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+ * "The feclearexcept() function shall attempt to clear the supported
+ * floating-point exceptions represented by excepts."
+ */
+
+#define _FPU_DEFAULT 0x0
+#define _FPU_IEEE 0x1F
+
+__declare_fenv_inline(int) feclearexcept(int excepts)
+{
+  unsigned int fcsr;
```
[the remark this answers is not in the document]
Comment (FlyGoat) [reply newlib/libc/machine/loongarch/machine/fenv-fp.h:50]: It’s always 32bit. Will try uint32_t.

File: newlib/libc/machine/loongarch/machine/fenv-fp.h [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+ * unspecified, except that if the excepts argument represents IEC
+ * 60559 valid coincident floating-point exceptions for atomic
+ * operations (namely overflow and inexact, or underflow and inexact),
+ * then overflow or underflow shall be raised before inexact. Whether
+ * the feraiseexcept() function additionally raises the inexact
+ * floating-point exception whenever it raises the overflow or
+ * underflow floating-point exception is implementation-defined."
+ */
+
+__declare_fenv_inline(int) feraiseexcept(int excepts)
```
[the remark this answers is not in the document]
Comment (FlyGoat) [reply newlib/libc/machine/loongarch/machine/fenv-fp.h:212]: I guess other libc do that with trapping in mind. Is it possible that RTOS using picolibc wants trapping?

[the remark this answers is not in the document]
Comment (keith-packard) [reply newlib/libc/machine/loongarch/machine/fenv-fp.h:212]: Yes, and that might be nice to have, so perhaps you should include this stuff. Then we can treat all of the other archs as faulty :-)

File: newlib/libc/machine/loongarch/machine/fenv-fp.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (keith-packard) [reply newlib/libc/machine/loongarch/machine/fenv-fp.h:50]: Yeah, reading the API specs, all of these functions use 'int' parameters anyways; perhaps it's better to just use 'int' everywhere and avoid a pile of unnecessary casts. int is always 32-bit on loongarch, I assume?

File: newlib/libc/machine/loongarch/machine/_fpmath.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (keith-packard) [reply newlib/libc/machine/loongarch/machine/_fpmath.h:40]: If the target is always LSB, then, yeah, would be nice to not have MSB confusing people (like me).

File: newlib/libc/machine/loongarch/machine/fenv-fp.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (FlyGoat) [reply newlib/libc/machine/loongarch/machine/fenv-fp.h:50]: That’s true from LoongArch ABI convention :-)

File: newlib/libc/machine/loongarch/machine/_fpmath.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (FlyGoat) [reply newlib/libc/machine/loongarch/machine/_fpmath.h:40]: I just double-checked current implementation and I think it's already LSB only?

[the remark this answers is not in the document]
Comment (keith-packard) [reply newlib/libc/machine/loongarch/machine/_fpmath.h:40]: sounds good.

File: newlib/libc/machine/loongarch/machine/fenv-fp.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (FlyGoat) [reply newlib/libc/machine/loongarch/machine/fenv-fp.h]: Opps, seems like something went wrong here, will investigate.

File: newlib/libc/machine/loongarch/machine/fenv-fp.h [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+ * the value of a rounding direction macro, the rounding direction is
+ * not changed."
+ */
+
+__declare_fenv_inline(int) fesetround(int round)
+{
+
+  unsigned int fcsr;
+
+  if ((round & ~FE_RMODE_MASK) != 0)
```
[the remark this answers is not in the document]
Comment (FlyGoat) [reply newlib/libc/machine/loongarch/machine/fenv-fp.h]: So this check is actually necessary. See MIPS's implementation.

[the remark this answers is not in the document]
Comment (keith-packard) [reply newlib/libc/machine/loongarch/machine/fenv-fp.h:367]: Arm and aarch64 also check for invalid parameters. Thanks!

File: newlib/libc/machine/loongarch/asm.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
+/* Copyright (c) 2024 Jiaxun Yang <jiaxun.yang@flygoat.com>
+
+   This copyrighted material is made available to anyone wishing to use,
```
Comment (keith-packard) [inline newlib/libc/machine/loongarch/asm.h:3]: Could you replace this license text with the text used in newlib/libc/machine/loongarch/machine/fenv.h? 'The FreeBSD license' isn't well defined.

**Review by keith-packard — CHANGES_REQUESTED** [review]
I was just doing a final pass over the PR before pushing and discovered that newlib/libc/machine/loongarch/asm.h has an ambiguous license. You've used the standard 3-clause BSD license in most of the other new code; can you apply that to this file? If not, can you replace the reference to opensource.org/licenses with the actual license text?

**Comment by FlyGoat** [comment]
> I was just doing a final pass over the PR before pushing and discovered that newlib/libc/machine/loongarch/asm.h has an ambiguous license. You've used the standard 3-clause BSD license in most of the other new code; can you apply that to this file? If not, can you replace the reference to opensource.org/licenses with the actual license text?

Done! It was a bind copy :-)

**Pull request merged**